File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

# Import data type
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class NumberCounterNode(Node): 
    def __init__(self):
        super().__init__("number_counter") 

        self.counter_ = 0

        self.subscriber_ = self.create_subscription(Int64, "number",self.callback_number, 10)
        
        # Create publisher obj
        self.publisher_ = self.create_publisher(Int64, "number_count", 10) # type, topic name, queue size

        self.server_ = self.create_service(
            SetBool,"reset_counter", self.callback_reset_counter) # Srv_type, name_of_srv, callback
        self.get_logger().info("Reset counter server has been started")

    def callback_reset_counter(self, request, response):
        if(request.data):
            self.counter_ = 0
            self.get_logger().info("Reset counter")
            response.success = True
        return response     ## Always return the response

    def callback_number(self, msg):
        
        self.counter_+=1

        # Create msg obj
        msg = Int64()
        # Add custom data
        msg.data = self.counter_
        # Output the data
        self.publisher_.publish(msg)
        self.get_logger().info(str(msg.data))


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

class NumberCounterNode : public rclcpp::Node 
{
public:
    NumberCounterNode() : Node("number_counter") 
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>("number", 10,
                      std::bind(&NumberCounterNode::callback_number, this, std::placeholders::_1));
        
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count",10);
        RCLCPP_INFO(this->get_logger(), "Number Counter has been started.");

        server_ = this->create_service<example_interfaces::srv::SetBool>(
            "reset_counter",
            std::bind(&NumberCounterNode::callbackResetNumberCount, this,std::placeholders::_1, std::placeholders::_2));
        RCLCPP_INFO(this->get_logger(), "Service server has been started");
    }

private:
    void callback_number(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        (void)msg;
        auto publish_msg = example_interfaces::msg::Int64();
        publish_msg.data = ++this->counter_;
        publisher_->publish(publish_msg);

        RCLCPP_INFO(this->get_logger(), "%ld", publish_msg.data);
    }

    void callbackResetNumberCount(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                            const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if(request->data)
        {
            this->counter_ = 0;
            response->success = true;
            RCLCPP_INFO(this->get_logger(), "Resetting the counter to zero");
        }
    }

    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;    

    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    int counter_=0;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();   // MODIFY NAME
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}